### ISSUE TITLE
Pull type info from run-time type instead of static type

### ISSUE THREAD

**Issue by Aloshi** [Aloshi/dukglue#3]
If you pass a base class-typed pointer (`Base*`) that points to an object of a derived type (`Derived*`) into dukglue, the JavaScript object ends up with the Base class's prototype instead of the derived type's prototype. Example:

```cpp
class Base {}
class Derived : public Base {
  void doThing();
}

Base* obj = new Derived();
// pass obj to javascript with dukglue (typed as Base*)... (really need a nice helper for this)
```

Then later, we try to use a derived method in JavaScript:

```js
obj.doThing();  // doesn't work, obj is of type Base
```

An error occurs, because we pulled the type information from the static type (`typeid(obj) == typeid(Base)`), and not the dynamic type (`typeid(*obj) == typeid(Derived)`). See [this example](http://cpp.sh/25zvm).

**Comment by ptarabbia** [comment]
Is that really unexpected though ? You passed the object as a Base, so Javascript only sees the Base methods. 

As long as it works if you have a virtual doThing() defined in Base, I would think it's sufficient.

**Comment by Aloshi** [comment]
I'm late to respond, but: as a C++ developer, I agree that this isn't really unexpected behavior. I wouldn't have noticed if it had not been pointed out to me (maybe this is a common pattern in the JS world). A more realistic example that multiple people have given me is a factory function defined in C++, but where returned objects still have their unique class methods available in JavaScript:

```cpp
// C++
Shape* makeShape(const std::string& name)
{
	if (name == "rectangle")
		return new Rectangle();
	else if (name == "circle")
		return new Circle();
	else
		return NULL;
}
```

```js
// JS
var rect = makeShape('rectangle');
rect.doRectangularThing();  // should work!
```

**Comment by Aloshi** [comment]
Fixed in commit 5d0d670d05997c838307b2823900de9d0fefafc2.
